### PR TITLE
[SPARK-24339][SQL]Add project for transform/map/reduce sql to prune column

### PR DESCRIPTION
**Pull request by xdcjie**
## What changes were proposed in this pull request?

Transform query do not have Project Node, query like:

` select transform(a, b) using 'func' from e` 

and it' logic plan is:

> 'ScriptTransformation['a, 'b], func, [key#0, value#0] 
        +- 'UnresolvedRelation `e` (PlanTest.scala:97)

so that it will scan all the column data of relation.

In this PR, I propose to add Project Node for transform query, so that it scan required data by prune columns, for above transform query,  it's logic plan will be:

> 'ScriptTransformation['a, 'b], func, [key#0, value#0] 
   +- 'Project ['a, 'b]    
        +- 'UnresolvedRelation `e`

it will scan only two column data of relation.

In summary, Add Project Node for transform query can reduce the time of scan and assemble data.(In our scenario, the relation(table) have 700 columns.)

## How was this patch tested?

Modify existing test ("transform query spec")

### CODE REVIEW

**Comment by maropu** [comment]
@gatorsmile Can you trigger this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 1]
```diff
@@ -338,6 +338,17 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
         // Add where.
         val withFilter = relation.optionalMap(where)(filter)
 
+        // Add project.
+        val namedExpressions = expressions.map {
+          case e: NamedExpression => e
+          case e: Expression => UnresolvedAlias(e)
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: nit: `case e: _ => UnresolvedAlias(e)`

**Comment by maropu** [comment]
Could you add explain result differences with/without this pr in the description?

**Comment by xdcjie** [comment]
@maropu  I updated the commet. In summary, with this pr can reduce the time of scan and assemble data. In our scenario, the relation(table) have 700 columns.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 1 above)
Comment (xdcjie) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: the type of expressions is Expression, so i think
case e: _ => UnresolvedAlias(e) and case e: Expression => UnresolvedAlias(e) is equivalent.
Did have other reasons to change this?

Comment (xdcjie) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: The style is updated, review this, please.

**Comment by xdcjie** [comment]
jenkins, test this, please

**Comment by xdcjie** [comment]
@maropu @gatorsmile Do you have any comment/suggestion for this PR? Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 1 above)
Comment (xdcjie) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: I tried, "case e: _ => UnresolveAlias(e)" will occur "unbound wildcard type" complie error. I will revert to "case e: Expression => UnresolvedAlias(e)".

**Comment by xdcjie** [comment]
@maropu @gatorsmile Can you help me review this pr ?  Thanks。 #

**Comment by maropu** [comment]
kindly pining @gatorsmile @HyukjinKwon @ueshin

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by xdcjie** [comment]
retest this please

**Comment by xdcjie** [comment]
retest this please

**Comment by xdcjie** [comment]
retest this please

**Comment by xdcjie** [comment]
@HyukjinKwon Our project is based on the branch-2.2, we will merge the patch to local branch manually if it is against master. we prefer following community to local branch.

 If you don't like this approach, I will close the PR and make a new PR to against master.

**Comment by HyukjinKwon** [comment]
Yea, usually we do it for master first and then backport it to other branches to reduce the diff and master has the fix. I would appreciate it if you go in this way.

**Comment by gatorsmile** [comment]
Sorry, the fix does not look good to me. We should let the optimizer add the project automatically.

**Comment by HyukjinKwon** [comment]
@xdcjie, mind closing this one?

**Comment by xdcjie** [comment]
Ok！

**Comment by gatorsmile** [comment]
@maropu Do you want to take this over and add such a project in `ColumnPruning`?

**Comment by xuanyuanking** [comment]
I want to give a follow up PR and cc @gatorsmile @maropu for a review.
